learn_evm/arithmetic-checks.md

```diff
@@ -0,0 +1,668 @@
+`0x7fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff` which is equal to half of the maximum value that can be represented using uint256.
+The most significant bit of this number is `0`, while all other bits are `1`.
+
+On the other hand, all negative numbers start with a `1` as their first bit.
+If we look at the underlying hex representation of these numbers, they are all greater than or equal to the smallest integer that can be represented using int256, which is `0x8000000000000000000000000000000000000000000000000000000000000000` (equal to `1` shifted 255 bits to the left).
+
+To obtain the negative value of an integer in a two's complement system, we can flip all the underlying bits and add `1`: `-a = ~a + 1`.
+An example illustrates this.
+
+```solidity
+0x0000000000000000000000000000000000000000000000000000000000000003 // 3
+0xfffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffc // ~3
+0xfffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffd // -3 = ~3 + 1
+```
+
+To verify that `-a + a = 0` holds for all integers, we can use the property of two's complement arithmetic that `-a = ~a + 1`.
+By substituting this into the equation, we get `-a + a = (~a + 1) + a = MAX + 1 = 0`, where `MAX` is the maximum integer value.
+
+There is a unique case that needs special attention in two's complement arithmetic.
+The smallest possible integer `int256).min = 0x8000000000000000000000000000000000000000000000000000000000000000 = -57896044618658097711785492504343953926634992332820282019728792003956564819968`
+does not have a positive inverse for addition, making it the only negative number with this property.
+
+Interestingly, if we try to compute `-type(int256).min`, we get the same number, as `-type(int256).min = ~type(int256).min + 1 = type(int256).min`.
+This means there are two fixed points for additive inverses: `-0 = 0` and `-type(int256).min = type(int256).min`.
+It's important to note that Solidity's arithmetic checks will throw an error when evaluating `-type(int256).min` (outside of unchecked blocks).
+
+Finally, looking at the underlying bit (or hex) representation highlights the importance of using the correct operators for signed integers, such as `slt` instead of `lt`, to avoid misinterpreting negative values as large numbers.
+
+```solidity
+  0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff // int256(-1) or type(uint256).max
+< 0x0000000000000000000000000000000000000000000000000000000000000000 // 0
+// When using `slt`, the comparison is interpreted as `-1 < 0 = true`.
+= 0x0000000000000000000000000000000000000000000000000000000000000001
+// When using `lt`, the comparison is interpreted as `type(uint256).max < 0 = false`.
+= 0x0000000000000000000000000000000000000000000000000000000000000000
+```
+
+Newer versions of Solidity prevent integer overflow by using the computed result `c = a + b` to check for overflow/underflow.
```

Comment:
What are the newer versions here? (0.8.16?)
Remind that this page will be still published when solidity 10 will be released ;)

Reply:
What do you mean by solidity 10? Solidity v0.10.0?

Would you always prefer to have the versions included in the text? I moved them to the code instead to de-clutter and to make it clearer where exactly they belong to.

Reply:
yes I meant 0.10.0. My point is that this text will still be here in 1 year, so it's better to avoid "new version"/"older version" etc, because it won't make sense later.

Adding the version in both the text and the code snippet is probably our best solution here